### PR TITLE
Feed db fix

### PR DESCRIPTION
**Pull request by allisterakun**
- Added a new entry `"nutrient_standard"` in the `"config.json"` file to provide the option for choosing between NRC or NASEM.
- Removed "animal_properties.animal_config.management_decisions.energy_and_nutrient_calculation_method" in all metadata JSON and animal module input JSON files.
- Replaced the `"energy_and_nutrient_calculation_method"` variable in `"AnimalBase.config"` (the AnimalBase class in `RUFAS.routines.animal.life_cycle.animal_base.py`) with `"nutrient_standard"`.
  - manually adding the `"nutrient_standard"` property into the `"config"` dictionary in the `"AnimalBase"` class (`RUFAS.routines.animal.life_cycle.animal_base.py line 37`)
  - replaced all references to `"AnimalBase.config["energy_and_nutrient_calculation_method"]"` with `"AnimalBase.config["nutrient_standard"]"
- Put together `"NASEM_comp_with_TDN.csv"` by inner joining NRC and NASEM on `NRC.rufas_id = NASEM.rufas_id` and adding the corresponding TDN column to each row.
- Removed database-related properties in all feed input JSON files and from the `feed_properties` section in all the metadata files: (`feed_database`, `feeds_table`, `feed_quality_table`, `nutrient_table`)
- Renamed columns in user_feed-related CSV files (took out the leading white space).

@JoeWaddell chiming in:

The temporary TDN and DE fixes are because the NRC (2001 nutrient requirements guidelines/book) and the NASEM (2021 nutrient requirements guidelines/book) use different methodologies not only for calculating animal requirements and feed supply, but they include different attributes for each feed item in the list (the nutrient composition list this PR is handling). The two libraries also don't overlap entirely with respect to which feeds are included. So we now have a singular RuFaS ID that identifies every feed, but some are only in NRC, some only in NASEM.

So for the temporary fixes: Total digestible nutrients (TDN) was essentially deprecated in NASEM, and so the NASEM feed library doesn't include those values for each feed. We have a temporary csv `NASEM_comp_with_TDN.csv` where @allisterakun removed any NASEM feed not in the NRC file, and added the TDN value from the NRC file.

Likewise, NRC assigned a single Digestible Energy (DE) value to each feed, and NASEM instead uses a DE_base value, which is then modified internally to calculate the 'actual' DE value.  So we've temporarily been using the DE_Base value - since it's *close* - until we add the DE calculations for the NASEM equations that require it.

Both will be addressed after Milestone 3, see Issue #834.

### CODE REVIEW

**Review by JoeWaddell — COMMENTED** [review]
There's still a discrepancy in nomenclature between the two feed databases: `DE_base` is equivalent to `DE`. If I try to run the simulation using `NRC` as the `nutrient_standard`, it fails because it can't find `DE_base`. I think better to rename DE_base in the feed database to DE, and revert the changes in Feed.py, but we can also now loop @Ed731551 into the conversation since he's back in the lab.

**Comment by Ed731551** [comment]
> There's still a discrepancy in nomenclature between the two feed databases: `DE_base` is equivalent to `DE`. If I try to run the simulation using `NRC` as the `nutrient_standard`, it fails because it can't find `DE_base`. I think better to rename DE_base in the feed database to DE, and revert the changes in Feed.py, but we can also now loop @Ed731551 into the conversation since he's back in the lab.

Thanks Joe. I agree with your suggestion to rename DE base in the feed database. The thing with NASEM (2021) is that there 'DE Base' value linked to every single feed in the database. To make NASEM values equivalent with the tabulated 'DE' values reported in NRC (2001) tables, we should implement adjustments to the base digestibilities depending on NDF and starch contents as explained in the energy chapter (pages 25 and 26). Chapter summary and formulas are presented in the RuFaS  pseudocodes for NASEM.

File: RUFAS/classes.py
Comment (Isgotkowitz) [inline RUFAS/classes.py]:
```suggestion
        nutrient_standard = im.get_data("config.nutrient_standard")
        self.feed = Feed(feed_class_config, nutrient_standard)
```
Not important, but keeps line consistent with above and below lines.

File: RUFAS/routines/animal/life_cycle/animal_base.py
Comment (Isgotkowitz) [inline RUFAS/routines/animal/life_cycle/animal_base.py:37]: Can a docstring be added here? This is a confusing method.
Is the config that is passed as an argument to `set_config()` the same as the config that can be pulled from the InputManager?

File: RUFAS/routines/animal/ration/animal_requirements.py
Comment (Isgotkowitz) [inline RUFAS/routines/animal/ration/animal_requirements.py]:
```suggestion
                {AnimalBase.config['nutrient_standard']} \
```
I think this ensures there is a space between the formatted input and "not supported", but not 100% sure.

File: RUFAS/routines/animal/ration/animal_requirements.py
Comment (Isgotkowitz) [inline RUFAS/routines/animal/ration/animal_requirements.py:595]: I think the name of the error (`"nutrient_standard_error"`) doesn't have to be in snake case. Other places in the code it isn't used that way, but there is no distinct standard for this.

File: RUFAS/routines/animal/ration/calf_ration.py
Comment (Isgotkowitz) [inline RUFAS/routines/animal/ration/calf_ration.py]:
```suggestion
    de_key = 'DE_Base' if 'DE_Base' in calf_feeds[whole_milk_id].keys() else 'DE'
    whole_milk_de = calf_feeds[whole_milk_id][de_key]
    milk_replacer_de = calf_feeds[milk_replacer_id][de_key]
    starter_de = calf_feeds[starter_id][de_key]
```

File: RUFAS/routines/animal/ration/calf_ration.py
Comment (Isgotkowitz) [inline RUFAS/routines/animal/ration/calf_ration.py]:
```suggestion
    me_intake = whole_milk_me * whole_milk_intake + milk_replacer_me * milk_replacer_intake + starter_me \
```

File: RUFAS/routines/animal/ration/ration_driver.py
Comment (Isgotkowitz) [inline RUFAS/routines/animal/ration/ration_driver.py:1033]: Is this an issue on GitHub? Is the plan to have it fixed soon?

File: RUFAS/routines/feed/feed.py
Comment (Isgotkowitz) [inline RUFAS/routines/feed/feed.py]:
```suggestion
    Pack the provided variable names and values into a dictionary.

    Parameters
    ----------
    var_names : List[str]
        List of keys for the resulting dictionary.
    var_values : List[Any]
        Corresponding values for the keys specified in `var_names`.

    Returns
    -------
    Dict[str, Any]
        Dictionary constructed from the provided variable names and values.

    Raises
    ------
    ValueError
        If the lengths of `var_names` and `var_values` are not the same.

    Notes
    -----
    Ensure that both `var_names` and `var_values` have the same length.
    Each name in `var_names` will be paired with its corresponding value from `var_values`.

    Example
    -------
    >>> pack_into_dict(['a', 'b'], [1, 2])
    {'a': 1, 'b': 2}

    """
```

File: RUFAS/routines/feed/feed.py
Comment (Isgotkowitz) [inline RUFAS/routines/feed/feed.py]:
```suggestion
    Parameters
    ----------
    data_source : str
        Path to retrieve the desired data from IM.
    var_names : List[str] = None
        A list of desired variables (columns) to retrieve.
        If omitted, all existing columns are returned.
    unique_value: bool = False
        If True, returns only unique values. Default is False.
    identifier: str = None
        Column name used for filtering based on `desired_rows`.
        Both `identifier` and `desired_rows` must be provided for filtering.
    desired_rows: List[Any] = None
        Desired row values for filtering. Rows with these values in the `identifier` column are returned.
    compare_val: int = None
        Baseline value for comparison. Used with `low_col` and `high_col` to filter rows. The "compare_val" should
        fall in the range of [low_col_value, high_col_value], otherwise the current row will be ignored.
    low_col: str = None
        Column indicating the lower bound for comparison with `compare_val`.
    high_col: str = None
        Column indicating the upper bound for comparison with `compare_val`.

    Returns
    -------
    List[Dict[str, Any]]
        Returns a list of dictionaries, each dictionary corresponds to a row in the csv file.
```

File: RUFAS/routines/feed/feed.py
Comment (Isgotkowitz) [inline RUFAS/routines/feed/feed.py:64]: Update docstring

File: RUFAS/routines/feed/feed.py
Comment (Isgotkowitz) [inline RUFAS/routines/feed/feed.py]:
```suggestion
```

File: RUFAS/routines/feed/feed.py
Comment (Isgotkowitz) [inline RUFAS/routines/feed/feed.py:1077]: Same as above, is there a plan for addressing these, is there an issue for them in GitHub?

File: RUFAS/scenario_manager.py
Comment (Isgotkowitz) [inline RUFAS/scenario_manager.py]: Revert changes before merging

File: input/data/feed/user_feeds_old.csv
Comment (Isgotkowitz) [inline input/data/feed/user_feeds_old.csv:1]: Is there a better name that could be used than `user_feeds_old.csv`? To me, it makes it sound like a deprecated version of `user_feeds.csv` even though it probably isn't.

File: input/metadata/ARL_metadata.json
Comment (Isgotkowitz) [inline input/metadata/ARL_metadata.json]:
```suggestion
        "pattern": "^(NASEM|NRC)$",
```
Might be a little overkill, but is what `"energy_and_nutrient_calculation_method"` had.

File: tests/test_animal_manager.py
Comment (Isgotkowitz) [inline tests/test_animal_manager.py]:
```suggestion
def animal_manager(input_manager: InputManager, mock_im_pool: Dict[str, Dict[str, Any]]) -> AnimalManager:
```

File: RUFAS/routines/feed/feed.py
Comment (Isgotkowitz) [inline RUFAS/routines/feed/feed.py]: Two questions on this method:
1. Can you put some examples in the docstring? I understand what this method does at a high level but I am having trouble comprehending how some of the different inputs effect output.
2. Does there need to be any error raising/validation in this method? What if two columns in the CSV are different lengths, what if unique_value is set to true but finds duplicate values, etc.

File: RUFAS/routines/feed/feed.py
Comment (Isgotkowitz) [inline RUFAS/routines/feed/feed.py:1]: `pack_into_dict()` and `retrieve_data()` need tests.

**Review by Isgotkowitz — CHANGES_REQUESTED** [review]
Looks good for the most part. Requesting changes because there appear to be some missing feed id's and I want to confirm there is a plan to find them or an issue exists for this, and also the new methods in `feed.py` are untested.

Everything else is small stuff, nice job Allister.

**Review by Isgotkowitz — APPROVED** [review]
Talked w/ Allister about missing values and tests, is plan in place for taking care of those.

File: RUFAS/routines/animal/ration/ration_driver.py
Comment (allisterakun) [reply RUFAS/routines/animal/ration/ration_driver.py:1033]: This is just a temporary fix, it will be taken care of when we refactor the feed module.

File: RUFAS/routines/feed/feed.py
Comment (allisterakun) [reply RUFAS/routines/feed/feed.py:1]: Will be taken care of when refactoring the feed module

File: RUFAS/routines/feed/feed.py
Comment (allisterakun) [reply RUFAS/routines/feed/feed.py:1077]: Will be taken care of when refactoring the feed module.

**Comment by ew3361zh** [comment]
This looks good overall. I think the other approval should probably come from @JoeWaddell or another SME instead of a second dev but one thing I'd appreciate is a little more context on what the overall issue(s) this PR is addressing. I see the solutions below but I'm not sure why. So maybe it need more of the why part of the what-why-how of the initial description.

> * Added a new entry `"nutrient_standard"` in the `"config.json"` file to provide the option for choosing between NRC or NASEM.
> * Removed "animal_properties.animal_config.management_decisions.energy_and_nutrient_calculation_method" in all metadata JSON and animal module input JSON files.

e.g. I'm not sure why this is a good thing to do or why it's needed.

> * Replaced the `"energy_and_nutrient_calculation_method"` variable in `"AnimalBase.config"` (the AnimalBase class in `RUFAS.routines.animal.life_cycle.animal_base.py`) with `"nutrient_standard"`.

same here.

>   * manually adding the `"nutrient_standard"` property into the `"config"` dictionary in the `"AnimalBase"` class (`RUFAS.routines.animal.life_cycle.animal_base.py line 37`)
>   * replaced all references to `"AnimalBase.config["energy_and_nutrient_calculation_method"]"` with `"AnimalBase.config["nutrient_standard"]"
> * Put together `"NASEM_comp_with_TDN.csv"` by inner joining NRC and NASEM on `NRC.rufas_id = NASEM.rufas_id` and adding the corresponding TDN column to each row.

Maybe even linking to an issue this is solving would be good.

File: RUFAS/routines/feed/feed.py
Comment (allisterakun) [reply RUFAS/routines/feed/feed.py]: 1. Added more examples.
2. Ideally, there shouldn't be any mismatches because all the data should be validated during the import process in IM. However, I moved the two functions (`retrieve_data()` and `pack_into_dict()`) into the Feed class to handle the possible length mismatch exception.

File: input/data/feed/user_feeds_old.csv
Comment (allisterakun) [reply input/data/feed/user_feeds_old.csv:1]: @JoeWaddell Can you please comment on this?

File: RUFAS/routines/animal/ration/animal_requirements.py
Comment (allisterakun) [reply RUFAS/routines/animal/ration/animal_requirements.py:595]: Yah I just sticked with the previous naming standard.

File: RUFAS/routines/animal/life_cycle/animal_base.py
Comment (allisterakun) [reply RUFAS/routines/animal/life_cycle/animal_base.py:37]: It looks like that the "config" here refers to im.get_data("animal.animal_config"). I do believe there are way too many different "config"s in this codebase, I will speak to @PooyaHekmati and make a separate Git Issue on this. But I don't believe this is the main focus of this PR,

File: RUFAS/routines/animal/ration/animal_requirements.py
Comment (allisterakun) [reply RUFAS/routines/animal/ration/animal_requirements.py]: Done

File: RUFAS/scenario_manager.py
Comment (allisterakun) [reply RUFAS/scenario_manager.py]: Done!

**Comment by JoeWaddell** [comment]
> This looks good overall. I think the other approval should probably come from @JoeWaddell or another SME instead of a second dev but one thing I'd appreciate is a little more context on what the overall issue(s) this PR is addressing. I see the solutions below but I'm not sure why. So maybe it need more of the why part of the what-why-how of the initial description.

I wasn't sure if I should approve it or not, since I made a few commits. I'll add to the what-why-how in the initial description to explain from the SME side why some of the temporary fixes are in, and I'll link to the appropriate Issue.

**Comment by allisterakun** [comment]
> > This looks good overall. I think the other approval should probably come from @JoeWaddell or another SME instead of a second dev but one thing I'd appreciate is a little more context on what the overall issue(s) this PR is addressing. I see the solutions below but I'm not sure why. So maybe it need more of the why part of the what-why-how of the initial description.
> 
> I wasn't sure if I should approve it or not, since I made a few commits. I'll add to the what-why-how in the initial description to explain from the SME side why some of the temporary fixes are in, and I'll link to the appropriate Issue.

Thank you! I will ask @tomhuhh for another review. Also looping in @KFosterReed for the update.

**Review by tomhuhh — APPROVED** [review]
LGTM! After further discussion with Joe on this, I think for the purpose of this PR, it is good to go. Good work Allister! Thanks!

**Comment by tomhuhh** [comment]
More notes from my discussion with Joe @JoeWaddell: 

Eventually - add NASEM equivalents to the constraint methods in ration optimizer.py and get rid of the NASEM_Comp_with_TDN table 
	When - at least after the animal manager refactor - another month or in January

**Pull request merged**
